hash_practice/exercises.py

```diff
@@ -25,5 +64,119 @@ def valid_sudoku(table):
+                return False
+            else:
+                seen_dict[elem] = 1
+        seen_dict = {}
+
+    # for vertical
+    seen_dict = {}
+    for col in range(len(table[0])):
+        for row in range(len(table)):
+            if table[row][col] == ".":
+                        continue
+            if table[row][col] in seen_dict:
+                    return False
+            else:
+                    seen_dict[table[row][col]] = 1
+        seen_dict = {}
+    
+    # # for 3x3 grid
+    seen_dict = {}
+    for i in range(3):
+        for j in range(3):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+
+    seen_dict = {}
+    for i in range(3):
+        for j in range(3, 6):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+
+    seen_dict = {}
+    for i in range(3):
+        for j in range(6,9):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+
+    seen_dict = {}
+    for i in range(3,6):
+        for j in range(3):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
 
+    seen_dict = {}
+    for i in range(3,6):
+        for j in range(3, 6):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+    
+    seen_dict = {}
+    for i in range(3,6):
+        for j in range(6, 9):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+    
+    seen_dict = {}
+    for i in range(6,9):
+        for j in range(3):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+    
+    seen_dict = {}
+    for i in range(6,9):
+        for j in range(3, 6):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1
+    
+    seen_dict = {}
+    for i in range(6,9):
+        for j in range(6, 9):
+            if table[i][j] == ".":
+                continue
+            if table[i][j] in seen_dict:
+                return False
+            else:
+                seen_dict[table[i][j]] = 1

```

Comment:
This works, but usually when you repeat almost the same code so many times, it's usually worth thinking about how you can cut down on duplication through perhaps defining a helper function.

Code repetition like this tends to lead to errors and higher maintenance as the different versions of almost the same code fall out of sync with each other.